hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala

```diff
@@ -429,108 +407,105 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
+    val srcAddrRef = mb.newLocal[Long]
+    val len = mb.newLocal[Int]
 
-    val dstAddress = mb.newField[Long]
-    val numberOfElements = mb.newLocal[Int]
-    val currentElementAddress = mb.newLocal[Long]
-    val currentIdx = mb.newLocal[Int]
+    Code(
+      srcAddrRef := srcAddress,
+      len := sourceType.loadLength(srcAddrRef),
+      constructOrCopyWithLen(mb, region, sourceType, srcAddrRef, len, forceDeep))
+  }
 
-    val init = Code(
-      numberOfElements := sourceType.loadLength(srcAddress),
-      dstAddress := this.allocate(region, numberOfElements),
-      this.stagedInitialize(dstAddress, numberOfElements),
-      currentElementAddress := this.firstElementOffset(dstAddress, numberOfElements),
-      currentIdx := const(0)
-    )
+  def copyFromType(region: Region, srcPType: PType, srcAddress: Long, forceDeep: Boolean): Long = {
+    val sourceType = srcPType.asInstanceOf[PArray]
+    val len = sourceType.loadLength(srcAddress)
+    constructOrCopyWithLen(region, sourceType, srcAddress, len, forceDeep)
+  }
 
-    var loop: Code[Unit] =
-      destElementType.storeShallowAtOffset(
-        currentElementAddress,
-        destElementType.copyFromType(
-          mb,
-          region,
-          sourceElementType,
-          sourceType.loadElement(srcAddress, numberOfElements, currentIdx),
-          forceDeep
-        )
-      )
+  def copyFromTypeAndStackValue(mb: MethodBuilder, region: Code[Region], srcPType: PType, stackValue: Code[_], forceDeep: Boolean): Code[_] =
+    this.copyFromType(mb, region, srcPType, stackValue.asInstanceOf[Code[Long]], forceDeep)
 
-    if(!sourceElementType.required) {
-      loop = sourceType.isElementMissing(srcAddress, currentIdx).mux(
-        this.setElementMissing(dstAddress, currentIdx),
-        loop
-      )
-    }
+  def constructAtAddress(mb: MethodBuilder, addr: Code[Long], region: Code[Region], srcPType: PType, srcAddress: Code[Long], forceDeep: Boolean): Code[Unit] = {
+    val srcArray = srcPType.asInstanceOf[PArray]
 
+    val srcAddrVar = mb.newLocal[Long]
+    val len = mb.newLocal[Int]
     Code(
-      init,
-      Code.whileLoop(currentIdx < numberOfElements,
-        loop,
-        currentElementAddress := this.nextElementAddress(currentElementAddress),
-        currentIdx := currentIdx + const(1)
-      ),
-      dstAddress
+      srcAddrVar := srcAddress,
+      len := srcArray.loadLength(srcAddrVar),
+      Region.storeAddress(addr, constructOrCopyWithLen(mb, region, srcArray, srcAddrVar, len, forceDeep))
     )
   }
 
-  def copyFromType(region: Region, srcPType: PType, srcAddress: Long, forceDeep: Boolean): Long = {
-    val sourceType = srcPType.asInstanceOf[PArray]
-    val sourceElementType = sourceType.elementType.fundamentalType
-    val destElementType = this.elementType.fundamentalType
+  def constructOrCopyWithLen(mb: MethodBuilder, region: Code[Region], srcArray: PArray, srcAddress: LocalRef[Long], len: LocalRef[Int], forceDeep: Boolean): Code[Long] = {
```

Comment:
to make sure I understand, what is the advantage of LocalRef vs FieldRef, which I think we could also use. I know we have had some issues with the scope of locals (making sure the local is defined in the scope it is used, which isn't a problem here), and I think the tradeoff vs FieldRef / newField is that newField adds a property on the class that MethodBuilder emits, and the JVM has a limit on the size of the class (64KB), and so we can run into class size issues. Is this precise enough, if not would you mind sharing the correct description?

Reply:
yes, that's how I understand the difference.

Reply:
I think these can be private

Reply:
oh, good call